_published/0012-authentication.md

```diff
@@ -0,0 +1,167 @@
+- Author:
+  {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation
+  issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation
+  status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Wharf should have authentication. Wharf currently has no authentication. This
+will outline a plan to implement [OpenID Connect (OICD)](https://openid.net/connect/)
+in wharf. This RFC will only outline authentication with OICD and not
+authorization and user/group
+management. A further plan to implement user groups and authorization is
+expected to be drawn up in a later document.
+
+## Motivation
+
+Some claim a good feature of secure infrastructure is access control. Currently,
+our only solution is network isolation. Being able to restrict access to varying
+degrees is a product goal to satisfy business needs.
+
+## Explanation
+
+Pre implementation stages
+
+1. OICD in any form demands full SSL security in all connections. In order to
+   avoid having to fill the development environment with --force-unsecure
+   bindings we should consider implementing certificate generation and importing
+   for dev environments.
+
+Implementations stages
+
+1. Docker compose setup for Hydra Proxy setup
+   - This involves integration with the currently existing DB.
+2. Hydra (Authorisation server) - Connect to IDP (Config for Azure AD iver.se)
+   - Config instructions documented [here](https://www.ory.sh/hydra/docs/reference/configuration)
+   - [OpenID-Configuration](https://login.microsoftonline.com/common/.well-known/openid-configuration)
+   -
+3. Wharf-web - configure and setup [angular-auth-oicd-client](https://github.com/damienbod/angular-auth-oidc-client)
+   - Implement login endpoint
+   - Implement consent endpoint
+   - Implement logout endpoint
+
+4. Wharf-api - Implement and test
+   - [Go SDK for ory Hydra](https://www.ory.sh/hydra/docs/sdk/go)
+   - Inspiration for [gin middleware here](https://github.com/janekolszak/gin-hydra/blob/master/ginhydra.go)
+     note this is outdated.
+
+7. Edit wharf-web services to use bearer tokens through angular middleware
+   configured through `agular-auth-oicd-client`
+
+8. Wharf-provider-* - Implement and test
+   - Likely identical to procedure for wharf-api.
+
+### Flows
+
+OpenID Connect - Flows
+A good guide on flows is available [here on auth0](https://auth0.com/docs/flows).
+Or for a more fun version checkout [this game by auth0](https://iverab-my.sharepoint.com/:i:/g/personal/fredrik_dyrvold_iver_se/Ef41bR71qeNOk-QWhCar0V4BZN8WqvM7iEf5g7ml9_s7Ew?e=ZoZRN6)
+
+### Identity provider
+
+We should connect to azure ad on our local deployments.
+
+We should not hav our own identity provider setup with wharf for now.
```

Comment:
> I think this is a bad idea. I'm a hard advocate for forcing the local dev environment to only rely on local resources.

The concept of running locally without external connectivity seems highly valuable. 

> We should look at some basic IDP that we can host alongside the docker-compose setup, so when you have the images downloaded then Wharf will not require any internet connection.

In the ory environment this would require the setup of a local idp. This local idp (when sourced from ory) is called [Ory-Kratos](https://www.ory.sh/kratos). If we do this then i think we should put off the requirement of using the AzureAD till a later RFC ex. after authorisation is in place.

If we were to create/switch on a bypass for the authentication system based on a developement flag we may be able to get a prototype off the ground faster with a smaller change footprint. This could allow for continuing work on prototyping integration with AzureAD while still retaining non-internet-connected development. 

The azureAD still seems like a requirement for this feature as that is where Iver is at.

> To me it feels silly to require all devs to setup something like a GitHub app just to log into the local dev Wharf instance, and it will get tedious or in our way fast.

If we use azure or github as an identity provider then there is no need to setup anything more. The remote endpoints are login.microsoft.com or github.com depending on the github/azureAD choice.

> I think that having one from the get-go should be a priority

I agree we should be able to develop and use the dev environments locally without connectivity to the internet. I disagree with the assumption that we should have a local idp. Some disadvantages of having a local idp include
##### Sprawl of user administration
In a production environment with a separate idp of our own we would get extra work in the form of having to conform to usder off&on boarding procedures. Using our remote idp this would not be an issue as we would be piggypbacking off exsisting such infrastructure, past and present.
#####  GDPR Storage of additional user specific info
We will likely need to store the name of a user even if we only use the authorisation server. 
If we start hosting an IDP of our own we will also need to store email or other contact information in order to deal with account recovery options.
##### Complexity
If we need a local idp setup for wharf then that should be a project of its own that would need to get completed before the authentication project can be started. While getting a prototype user store up and running with ory kratos may be relativly simple- scaling that to production does not seem trivial. 
IMO Adding a requirement of a yet non-exsisting, non-planned idp would act as a full hard block/stop for this project.

Reply:
Seems like I don't know the full implications of the term IDP.

I meant that for local development only we have some local username+password store that perhaps the ORY Hydra talks to. Then for production we only rely on third-party IDP such as Azure AD.

By having a local dev-only solution then we could really cut some corners and use more or less anything that's OpenID compliant. Could be a simple page with a single hardcoded username+password, or even just a button that says "Yep, log me in to local dev Wharf" without even having to provide any credentials. Doesn't even have to be a "certified" solution, as long as it complies with the spec so that we can manually e2e test it.

My intention is not to include some self-hosted IDP when we deploy Wharf to k8s, as that bring up so many of the problems that you talked about. Better to have this auth opt-in until we get that Azure AD integration set up.